File: sdk/ruby/lib/AgentMailer/drafts/types/draft_update.rb
# frozen_string_literal: true

module AgentMailer
  module Drafts
    module Types
      class DraftUpdate < Internal::Types::Model
        field :inbox_id, -> { String }, optional: false, nullable: false, api_name: "inboxId"

        field :draft_id, -> { String }, optional: false, nullable: false, api_name: "draftId"

        field :to, -> { Internal::Types::Array[AgentMailer::Drafts::Types::DraftUpdateToItem] }, optional: true, nullable: false

        field :cc, -> { Internal::Types::Array[AgentMailer::Drafts::Types::DraftUpdateCcItem] }, optional: true, nullable: false

        field :bcc, -> { Internal::Types::Array[AgentMailer::Drafts::Types::DraftUpdateBccItem] }, optional: true, nullable: false

        field :reply_to, -> { Internal::Types::Array[AgentMailer::Drafts::Types::DraftUpdateReplyToItem] }, optional: true, nullable: false, api_name: "replyTo"

        field :subject, -> { String }, optional: true, nullable: false

        field :text, -> { String }, optional: true, nullable: false

        field :html, -> { String }, optional: true, nullable: false

        field :headers, -> { Internal::Types::Hash[String, String] }, optional: true, nullable: false

        field :attachments, -> { Internal::Types::Array[AgentMailer::Drafts::Types::DraftUpdateAttachmentsItem] }, optional: true, nullable: false

        field :reply_to_message_id, -> { String }, optional: true, nullable: false, api_name: "replyToMessageId"

        field :forward_message_id, -> { String }, optional: true, nullable: false, api_name: "forwardMessageId"

        field :send_at, -> { String }, optional: true, nullable: false, api_name: "sendAt"
      end
    end
  end
end

File: sdk/rust/src/api/types/draft_update.rs
pub use crate::prelude::*;

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq)]
pub struct DraftUpdate {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub to: Option<Vec<DraftUpdateToItem>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cc: Option<Vec<DraftUpdateCcItem>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub bcc: Option<Vec<DraftUpdateBccItem>>,
    #[serde(rename = "replyTo")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reply_to: Option<Vec<DraftUpdateReplyToItem>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub subject: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub html: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub headers: Option<HashMap<String, String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attachments: Option<Vec<DraftUpdateAttachmentsItem>>,
    #[serde(rename = "replyToMessageId")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reply_to_message_id: Option<String>,
    #[serde(rename = "forwardMessageId")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub forward_message_id: Option<String>,
    #[serde(rename = "sendAt")]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub send_at: Option<DateTime<FixedOffset>>,
}

impl DraftUpdate {
    pub fn builder() -> DraftUpdateBuilder {
        <DraftUpdateBuilder as Default>::default()
    }
}

#[derive(Clone, PartialEq, Default, Debug)]
#[non_exhaustive]
pub struct DraftUpdateBuilder {
    to: Option<Vec<DraftUpdateToItem>>,
    cc: Option<Vec<DraftUpdateCcItem>>,
    bcc: Option<Vec<DraftUpdateBccItem>>,
    reply_to: Option<Vec<DraftUpdateReplyToItem>>,
    subject: Option<String>,
    text: Option<String>,
    html: Option<String>,
    headers: Option<HashMap<String, String>>,
    attachments: Option<Vec<DraftUpdateAttachmentsItem>>,
    reply_to_message_id: Option<String>,
    forward_message_id: Option<String>,
    send_at: Option<DateTime<FixedOffset>>,
}

impl DraftUpdateBuilder {
    pub fn to(mut self, value: Vec<DraftUpdateToItem>) -> Self {
        self.to = Some(value);
        self
    }

    pub fn cc(mut self, value: Vec<DraftUpdateCcItem>) -> Self {
        self.cc = Some(value);
        self
    }

    pub fn bcc(mut self, value: Vec<DraftUpdateBccItem>) -> Self {
        self.bcc = Some(value);
        self
    }

    pub fn reply_to(mut self, value: Vec<DraftUpdateReplyToItem>) -> Self {
        self.reply_to = Some(value);
        self
    }

    pub fn subject(mut self, value: impl Into<String>) -> Self {
        self.subject = Some(value.into());
        self
    }

    pub fn text(mut self, value: impl Into<String>) -> Self {
        self.text = Some(value.into());
        self
    }

    pub fn html(mut self, value: impl Into<String>) -> Self {
        self.html = Some(value.into());
        self
    }

    pub fn headers(mut self, value: HashMap<String, String>) -> Self {
        self.headers = Some(value);
        self
    }

    pub fn attachments(mut self, value: Vec<DraftUpdateAttachmentsItem>) -> Self {
        self.attachments = Some(value);
        self
    }

    pub fn reply_to_message_id(mut self, value: impl Into<String>) -> Self {
        self.reply_to_message_id = Some(value.into());
        self
    }

    pub fn forward_message_id(mut self, value: impl Into<String>) -> Self {
        self.forward_message_id = Some(value.into());
        self
    }

    pub fn send_at(mut self, value: DateTime<FixedOffset>) -> Self {
        self.send_at = Some(value);
        self
    }

    /// Consumes the builder and constructs a [`DraftUpdate`].
    pub fn build(self) -> Result<DraftUpdate, BuildError> {
        Ok(DraftUpdate {
            to: self.to,
            cc: self.cc,
            bcc: self.bcc,
            reply_to: self.reply_to,
            subject: self.subject,
            text: self.text,
            html: self.html,
            headers: self.headers,
            attachments: self.attachments,
            reply_to_message_id: self.reply_to_message_id,
            forward_message_id: self.forward_message_id,
            send_at: self.send_at,
        })
    }
}
